experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+//! interfaces hard to formulate.  I welcome a correction here if I'm wrong.  The objective is to
+//! unify as many as possible of the above use cases in a trait that can express all of them.
+//!
+//! ### "Stringly typed" API
+//!
+//! One possibility is to require a string-like type (`AsRef<str>`).
+//!
+//! Pros:
+//!
+//! * This is the closest to the JavaScript API surface
+//! * It is the easiest one to implement.
+//!
+//! Cons:
+//!
+//! * It has the loosest guarantees as to the functionality it provides.
+//!
+//! Here's an example of how a stringly typed trait can be used.  We can specify a general
+//! constraint on an input parameter and get a functional API. In the example above it is
+//! `AsRef<str>`, but we could have it be `AsRef<T>` for some other suitably chosen trait `T`.
+//!
+//! ```
+//! /// This is the general trait.
+//! use std::convert::AsRef;
+//! pub trait Trait {
+//!     fn example(l: impl AsRef<str>);
+//! }
+//!
+//! /// Sample trait implementation.  We see below that two implementations
+//! /// can coexist.
+//! struct I{}
+//! impl Trait for I {
+//!     fn example(l: impl AsRef<str>) {
+//!         let s: &str = l.as_ref();
+//!         println!("result = {}", s);
+//!     }
+//! }
+//!
+//! /// One implementation (`&'static str`) and example use.
+//! let s: &'static str = "hello!";
+//! I::example("hello!");
+//!
+//! /// Another implementation (`struct J`) and example use.
+//! struct J {
+//!   rep: String,
+//! }
+//! impl AsRef<str> for J {
+//!     fn as_ref(&self) -> &str {
+//!         &self.rep
+//!     }
+//! }
+//! let j = J{ rep: "world!".to_string() };
+//! I::example(&j);
+//! ```
+//!
+//! ### Fallibility
+//! 
+//! The issue of fallibility in the API comes up because an implementor can decide to implement
+//! lazy checking of the constructed collaborator types.  A `LocaleIdentifier` type is one of
+//! those types.  While [`Locale`][lid] validates its inputs, [`rust_icu_uloc::ULoc`][uloc] does
+//! not: in fact, almost all of its operations are fallible, and there does not seem to be a way to
+//! validate `ULoc` eagerly, the way the underlying ICU4C API is defined.
+//!
+//!   [lid]: https://github.com/unicode-org/icu4x/pull/47
+//!   [uloc]: https://docs.rs/rust_icu_uloc/0.2.2/rust_icu_uloc/struct.ULoc.html
+//!
+//! Now since `Intl.supportedLocalesOf()` exists, we can be reasonably sure that any user will have
+//! the chance to obtain a valid locale either by taking one from the list of supported locales,
+//! or by language-matching desired locale with the list of supported locales ahead of time.
+//!
+//! This means that an infallible API would work for the case of `Intl.ListFormat`.  A fallible
+//! locale type must be adapted into an infallible type before the list formatter is constructed.
+//! This means that the implementor should be able to short-circuit validation errors in the 
+//! implementation if needed.
```

Comment:
In general, I don't believe we should be exposing infallible APIs for falliable operations. Saying that "we hope the user used supportedLocalesOf" is a very weak guarantee and generally leads to a bad API.

And ECMA402 *is* fallible. `new Intl.DateTimeFormat("en", {"month": 2})` throws.
We should be able to throw too.

Reply:
Ack, will do.

Reply:
Done.